compiler/src/dotty/tools/dotc/transform/SpecializeFunctions.scala

```diff
@@ -0,0 +1,143 @@
+package dotty.tools.dotc
+package transform
+
+import ast.Trees._, ast.tpd, core._
+import Contexts.Context, Types._, Decorators._, Symbols._, DenotTransformers._
+import SymDenotations._, Scopes._, StdNames._, NameOps._, Names._
+import MegaPhase.MiniPhase
+
+import scala.collection.mutable
+
+/** Specializes classes that inherit from `FunctionN` where there exists a
+ *  specialized form.
+ */
+class SpecializeFunctions extends MiniPhase with InfoTransformer {
+  import ast.tpd._
+  val phaseName = "specializeFunctions"
+  override def runsAfter = Set(ElimByName.name)
+
+  /** Transforms the type to include decls for specialized applys  */
+  override def transformInfo(tp: Type, sym: Symbol)(using Context) = tp match {
+    case tp: ClassInfo if !sym.is(Flags.Package) =>
```

Comment:
Something to take into account for performance: currently this code will be run on every single symbol that is referenced by the compiler, even if that symbol comes from the classpath, but we only care about adding the specialized apply decls to the classes we are currently compiling, so you should be able to do:
```suggestion
  /** Transforms the type to include decls for specialized applys  */
  override def transformInfo(tp: Type, sym: Symbol)(using Context) = tp match {
    case tp: ClassInfo if !sym.is(Flags.Package) && sym.isDefinedInCurrentRun =>
```

Reply:
Thanks for the tip @smarter 👍

Reply:
Even better: since this is an InfoTransformer you can override `infoMayChange` to filter out early everything that doesn't need to be forced: https://github.com/lampepfl/dotty/blob/d084b8bb7833e18650797b32bbb7fe05b35538bd/compiler/src/dotty/tools/dotc/core/DenotTransformers.scala#L53-L57

Reply:
It's done in 1eb7b5d 👍